File: backend/app/security.py
from __future__ import annotations

import hashlib
import hmac
import secrets
from typing import Tuple


def hash_password(password: str, salt: bytes | None = None) -> Tuple[str, str]:
    if salt is None:
        salt = secrets.token_bytes(16)
    digest = hashlib.pbkdf2_hmac(
        "sha256",
        password.encode("utf-8"),
        salt,
        120_000,
    )
    return salt.hex(), digest.hex()


def verify_password(password: str, salt_hex: str, digest_hex: str) -> bool:
    salt = bytes.fromhex(salt_hex)
    digest = hashlib.pbkdf2_hmac(
        "sha256",
        password.encode("utf-8"),
        salt,
        120_000,
    ).hex()
    return hmac.compare_digest(digest, digest_hex)


def generate_api_key() -> str:
    return secrets.token_urlsafe(32)

File: server/security.js
import crypto from "crypto";

export const hashPassword = (password, saltHex) => {
  const salt = saltHex ? Buffer.from(saltHex, "hex") : crypto.randomBytes(16);
  const hash = crypto.pbkdf2Sync(password, salt, 120000, 32, "sha256");
  return {
    salt: salt.toString("hex"),
    hash: hash.toString("hex"),
  };
};

export const verifyPassword = (password, saltHex, hashHex) => {
  const { hash } = hashPassword(password, saltHex);
  return crypto.timingSafeEqual(
    Buffer.from(hash, "hex"),
    Buffer.from(hashHex, "hex")
  );
};

export const generateApiKey = () =>
  crypto.randomBytes(32).toString("base64url");
